src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java

```diff
@@ -187,57 +202,159 @@ private void dataCmd() {
-      mailMessage = encoder.encode();
-      mailResult.setMessageID(encoder.getMessageID());
+  private void sendDataOfPart(EncodedPart part, Handler<Void> endHandler) {
+    if (isMultiPart(part)) {
+      sendMailHeaders(part);
+      sendMultiPart(part, 0, endHandler);
+    } else {
+      sendRegularPart(part, endHandler);
+    }
+  }
+
+  private void sendMultiPart(EncodedPart part, int i, Handler<Void> endHandler) {
+    // write boundary start
+    final StringBuilder sb = new StringBuilder("--").append(part.boundary());
+    connection.writeLine(sb.toString(), wrottern.getAndAdd(sb.length()) < 1000);
+    EncodedPart thePart = part.parts().get(i);
+
+    Handler<Void> nextHandler;
+    if (i == part.parts().size() - 1) {
+      // this is the last part
+      nextHandler = v -> {
+        connection.writeLine(sb.append("--").toString(), wrottern.getAndAdd(sb.length()) < 1000);
+        endHandler.handle(null);
+      };
+    } else {
+      // next part to do
+      nextHandler = v -> sendMultiPart(part, i + 1, endHandler);
+    }
+    if (isMultiPart(thePart)) {
+      nextHandler.handle(null);
+    } else {
+      // send single part with the endHandler
+      sendRegularPart(thePart, nextHandler);
+    }
+  }
+
+  private boolean isMultiPart(EncodedPart part) {
+    return part.parts() != null && part.parts().size() > 0;
+  }
+
+  private void sendMailHeaders(EncodedPart part) {
+    for (Map.Entry<String, String> entry: part.headers()) {
+      connection.writeLine(entry.toString(), wrottern.getAndAdd(entry.toString().length()) < 1000);
+    }
+    // send empty line between headers and body
+    connection.writeLine("", wrottern.get() < 1000);
+  }
+
+  private void sendRegularPart(EncodedPart part, Handler<Void> endHandler) {
+    sendMailHeaders(part);
+    if (part.body() != null) {
+      // send body string
+      for (String bodyLine: part.body().split("\n")) {
+        String line = bodyLine.startsWith(".") ? "." + bodyLine : bodyLine;
+        connection.writeLine(line, wrottern.getAndAdd(line.length()) < 1000);
+      }
+      endHandler.handle(null);
+    } else if (part.bodyStream() != null) {
+      BodyReadStream bodyReadStream = new BodyReadStream(part.bodyStream());
+      bodyReadStream.pipe().endOnComplete(false).to(connection.getSocket(), r -> {
+        if (r.succeeded()) {
+          endHandler.handle(null);
+        } else {
+          log.warn("Failed to transfer stream", r.cause());
+          handleError(r.cause());
+        }
+      });
+    } else {
+      throw new IllegalStateException("No mail body and stream found");
+    }
+  }
+
+  private Handler<Void> handlerInContext(Handler<Void> handler) {
+    return vv -> connection.getContext().runOnContext(v -> handler.handle(null));
+  }
+
+  // what we need: strings line by line with CRLF as line terminator
+  private class BodyReadStream implements ReadStream<Buffer> {
+
+    private final ReadStream<Buffer> stream;
+    private final int size = 57;
```

Comment:
what is this 57 magic number ?

Reply:
57 / 3 * 4 = 76. 76 + "\r\n".length() = 78. which is the most common used length for email attachment. See: https://tools.ietf.org/html/rfc5322#section-2.1.1

Use this size for each byte[] --> base64String conversion to make sure to get limited length string.

Reply:
can you add a comment explaining this ?